File: backend/app/domain/services/security_analysis_service.py
# Security analysis service
from typing import List, Dict, Any, Optional
from datetime import datetime, timedelta
from ..models.security_event import SecurityEvent, AttackType, SeverityLevel

class SecurityAnalysisService:
    
    @staticmethod
    def calculate_threat_level(events: List[SecurityEvent]) -> Dict[str, Any]:
        """Calculate threat level based on events"""
        if not events:
            return {
                "level": "low",
                "score": 0,
                "reasons": []
            }
        
        reasons = []
        score = 0
        
        # Check attack rate
        attack_count = sum(1 for e in events if e.is_attack)
        attack_rate = attack_count / len(events) if events else 0
        
        if attack_rate > 0.5:
            score += 40
            reasons.append("High attack rate detected")
        elif attack_rate > 0.2:
            score += 20
            reasons.append("Moderate attack rate")
        
        # Check critical events
        critical_count = sum(1 for e in events if e.severity == SeverityLevel.CRITICAL)
        if critical_count > 0:
            score += critical_count * 10
            reasons.append(f"{critical_count} critical events detected")
        
        # Check blocked rate
        blocked_count = sum(1 for e in events if e.is_blocked)
        block_rate = blocked_count / len(events) if events else 0
        if block_rate > 0.7:
            score += 30
            reasons.append("High block rate indicates active threats")
        
        # Determine level
        if score >= 70:
            level = "critical"
        elif score >= 50:
            level = "high"
        elif score >= 30:
            level = "medium"
        else:
            level = "low"
        
        return {
            "level": level,
            "score": min(100, score),
            "reasons": reasons
        }
    
    @staticmethod
    def identify_suspicious_ips(events: List[SecurityEvent]) -> List[Dict[str, Any]]:
        """Identify suspicious IPs from events"""
        ip_stats = {}
        
        for event in events:
            if event.source_ip not in ip_stats:
                ip_stats[event.source_ip] = {
                    "total_requests": 0,
                    "attack_count": 0,
                    "blocked_count": 0,
                    "attack_types": set()
                }
            
            stats = ip_stats[event.source_ip]
            stats["total_requests"] += 1
            
            if event.is_attack:
                stats["attack_count"] += 1
                if event.attack_type:
                    stats["attack_types"].add(event.attack_type)
            
            if event.is_blocked:
                stats["blocked_count"] += 1
        
        # Calculate threat scores
        suspicious_ips = []
        for ip, stats in ip_stats.items():
            threat_score = 0
            
            # Attack rate factor
            attack_rate = stats["attack_count"] / stats["total_requests"]
            threat_score += attack_rate * 40
            
            # Attack diversity factor  
            threat_score += len(stats["attack_types"]) * 10
            
            # Block rate factor
            block_rate = stats["blocked_count"] / stats["total_requests"]
            threat_score += block_rate * 30
            
            # Volume factor
            if stats["total_requests"] > 100:
                threat_score += 20
            elif stats["total_requests"] > 50:
                threat_score += 10
            
            if threat_score > 30:
                suspicious_ips.append({
                    "ip": ip,
                    "threat_score": min(100, threat_score),
                    "total_requests": stats["total_requests"],
                    "attack_count": stats["attack_count"],
                    "attack_types": list(stats["attack_types"])
                })
        
        return sorted(suspicious_ips, key=lambda x: x["threat_score"], reverse=True)
    
    @staticmethod
    def calculate_risk_score(event: SecurityEvent) -> float:
        """Calculate risk score for a single event"""
        score = 0.0
        
        # Severity factor
        severity_scores = {
            SeverityLevel.LOW: 10,
            SeverityLevel.MEDIUM: 30,
            SeverityLevel.HIGH: 60,
            SeverityLevel.CRITICAL: 100
        }
        score += severity_scores.get(event.severity, 0) * 0.4
        
        # Attack type factor
        if event.is_attack:
            score += 30
            
        # Anomaly score factor
        if event.anomaly_score > 10:
            score += 20
        elif event.anomaly_score > 5:
            score += 10
        
        # Blocked factor (reduces risk if blocked)
        if event.is_blocked:
            score *= 0.7
        
        return min(100.0, score)

File: frontend/src/domain/services/SecurityAnalysisService.ts
// Security analysis domain service
import { SecurityEvent, SecurityEventEntity, AttackType, SeverityLevel } from '../entities/SecurityEvent';
import { IpAddress } from '../value-objects/IpAddress';
import { TimeRange } from '../value-objects/TimeRange';

export interface ThreatLevel {
  level: 'low' | 'medium' | 'high' | 'critical';
  score: number;
  reasons: string[];
}

export interface AttackPattern {
  type: AttackType;
  frequency: number;
  timeRange: TimeRange;
  sourceIps: string[];
}

/**
 * Security analysis service for client-side analysis of security events.
 * Note: For dashboard statistics, prefer using pre-aggregated data from backend
 * to avoid fetching large amounts of individual events.
 */
export class SecurityAnalysisService {
  // Helper functions for SecurityEvent interface
  private isAttack(event: SecurityEvent): boolean {
    return event.attackType !== undefined;
  }

  private isCritical(event: SecurityEvent): boolean {
    return event.severity === SeverityLevel.CRITICAL;
  }

  private getSeverityScore(event: SecurityEvent): number {
    const scores = {
      [SeverityLevel.LOW]: 1,
      [SeverityLevel.MEDIUM]: 2,
      [SeverityLevel.HIGH]: 3,
      [SeverityLevel.CRITICAL]: 4
    };
    return scores[event.severity];
  }
  
  /**
   * Analyze threat level from individual events.
   * @deprecated Use backend aggregated statistics for better performance
   */
  analyzeThreatLevel(events: SecurityEvent[]): ThreatLevel {
    if (events.length === 0) {
      return { level: 'low', score: 0, reasons: [] };
    }

    const reasons: string[] = [];
    let score = 0;

    // Check attack frequency
    const attackCount = events.filter(e => this.isAttack(e)).length;
    const attackRate = attackCount / events.length;
    
    if (attackRate > 0.5) {
      score += 40;
      reasons.push('High attack rate detected');
    } else if (attackRate > 0.2) {
      score += 20;
      reasons.push('Moderate attack rate');
    }

    // Check critical events
    const criticalCount = events.filter(e => this.isCritical(e)).length;
    if (criticalCount > 0) {
      score += criticalCount * 10;
      reasons.push(`${criticalCount} critical events detected`);
    }

    // Check blocked rate
    const blockedCount = events.filter(e => e.blocked).length;
    const blockRate = blockedCount / events.length;
    if (blockRate > 0.7) {
      score += 30;
      reasons.push('High block rate indicates active threats');
    }

    // Determine level based on score
    let level: ThreatLevel['level'];
    if (score >= 70) level = 'critical';
    else if (score >= 50) level = 'high';
    else if (score >= 30) level = 'medium';
    else level = 'low';

    return { level, score, reasons };
  }

  detectAttackPatterns(events: SecurityEvent[]): AttackPattern[] {
    const patterns = new Map<AttackType, AttackPattern>();

    events
      .filter(e => e.attackType)
      .forEach(event => {
        const type = event.attackType!;
        
        if (!patterns.has(type)) {
          patterns.set(type, {
            type,
            frequency: 0,
            timeRange: new TimeRange(event.timestamp, event.timestamp),
            sourceIps: []
          });
        }

        const pattern = patterns.get(type)!;
        pattern.frequency++;
        
        if (!pattern.sourceIps.includes(event.sourceIp)) {
          pattern.sourceIps.push(event.sourceIp);
        }

        // Update time range
        if (event.timestamp < pattern.timeRange.startDate) {
          pattern.timeRange = new TimeRange(event.timestamp, pattern.timeRange.endDate);
        }
        if (event.timestamp > pattern.timeRange.endDate) {
          pattern.timeRange = new TimeRange(pattern.timeRange.startDate, event.timestamp);
        }
      });

    return Array.from(patterns.values())
      .sort((a, b) => b.frequency - a.frequency);
  }

  identifySuspiciousIps(events: SecurityEvent[]): Array<{ ip: string; threatScore: number; attackTypes: AttackType[] }> {
    const ipStats = new Map<string, { 
      attackCount: number; 
      totalCount: number; 
      attackTypes: Set<AttackType>;
      blockedCount: number;
    }>();

    events.forEach(event => {
      if (!ipStats.has(event.sourceIp)) {
        ipStats.set(event.sourceIp, {
          attackCount: 0,
          totalCount: 0,
          attackTypes: new Set(),
          blockedCount: 0
        });
      }

      const stats = ipStats.get(event.sourceIp)!;
      stats.totalCount++;
      
      if (this.isAttack(event)) {
        stats.attackCount++;
        if (event.attackType) {
          stats.attackTypes.add(event.attackType);
        }
      }
      
      if (event.blocked) {
        stats.blockedCount++;
      }
    });

    return Array.from(ipStats.entries())
      .map(([ip, stats]) => {
        // Calculate threat score for IP
        let threatScore = 0;
        
        // Attack rate factor
        const attackRate = stats.attackCount / stats.totalCount;
        threatScore += attackRate * 40;
        
        // Attack diversity factor
        threatScore += stats.attackTypes.size * 10;
        
        // Block rate factor
        const blockRate = stats.blockedCount / stats.totalCount;
        threatScore += blockRate * 30;
        
        // Volume factor
        if (stats.totalCount > 100) threatScore += 20;
        else if (stats.totalCount > 50) threatScore += 10;

        return {
          ip,
          threatScore: Math.min(100, threatScore),
          attackTypes: Array.from(stats.attackTypes)
        };
      })
      .filter(item => item.threatScore > 30)
      .sort((a, b) => b.threatScore - a.threatScore);
  }

  /**
   * Calculate risk score from individual events.
   * @deprecated Use backend aggregated statistics for better performance
   */
  calculateRiskScore(events: SecurityEvent[]): number {
    if (events.length === 0) return 0;

    let totalScore = 0;
    
    events.forEach(event => {
      let eventScore = this.getSeverityScore(event) * 10;
      
      if (this.isAttack(event)) eventScore *= 1.5;
      if (event.blocked) eventScore *= 0.7; // Lower risk if blocked
      
      totalScore += eventScore;
    });

    return Math.min(100, totalScore / events.length);
  }
}